### ISSUE TITLE
coverage is highlighting the tests

### ISSUE THREAD

**Issue by zenspider** [greghendershott/racket-mode#99]
I checked this against drracket's (much uglier) highlighting and they're not equivalent:

Repro, run racket-mode's coverage analysis vs drracket's on the following:

```racket
#lang racket/base

(require rackunit)
;; (require "lib/shared.rkt")

(module+ test
  (require rackunit))

(define member?                         ; from ch12.scm
  (lambda (a lat)
    (letrec ((yes? (lambda (l)
                     (cond [(null? l) #f]
                           [(eq? (car l) a) #t]
                           [else (yes? (cdr l))]))))
      (yes? lat))))

(module+ test
  (test-case "member?"
    (check-false (member? 42 '(a b c)))
    (check-true (member? 'b '(a b c)))))

(define intersect1
  (lambda (set1 set2)
    (cond [(null? set1) '()]
          [(member? (car set1) set2)
           (cons (car set1)
                 (intersect1 (cdr set1) set2))]
          [else  (intersect1 (cdr set1) set2)])))

(define intersect2
  (lambda (set1 set2)
    (letrec ((I (lambda (set)
                  (cond [(null? set) '()]
                        [(member? (car set) set2)
                         (cons (car set)
                               (I (cdr set)))]
                        [else  (I (cdr set))]))))
      (I set1))))

(define intersectall1
  (lambda (lset)
    (letrec ((I (lambda (lset)
                  (cond [(null? (cdr lset)) (car lset)]
                        [else (intersect2 (car lset) (I (cdr lset)))]))))
      (cond ((null? lset) '())
            (else (I lset))))))

(define intersectall2
  (lambda (lset)
    (let/cc hop
            (letrec
                ((I (lambda (lset)
                      (cond [(null? (car lset)) (hop '())]
                            [(null? (cdr lset)) (car lset)]
                            [else (intersect2 (car lset) (I (cdr lset)))]))))
              (cond ((null? lset) '())
                    (else (I lset)))))))

(module+ test
  (check-equal? (intersectall2 '((a b c) (b c) (b)))
                '(b)))

(define intersect3
  (lambda (set1 set2)
    (letrec ((I (lambda (set)
                  (cond [(null? set) '()]
                        [(member? (car set) set2)
                         (cons (car set)
                               (I (cdr set)))]
                        [else  (I (cdr set))]))))
      (cond
       ((null? set2) '())
       (else (I set1))))))

(define intersectall3
  (lambda (lset)
    (let/cc hop
            (letrec
                ((A (lambda (lset)
                      (cond [(null? (car lset)) (hop '())]
                            [(null? (cdr lset)) (car lset)]
                            [else (intersect3 (car lset) (A (cdr lset)))])))
                 (I (lambda (s1 s2)
                      (letrec ((J (lambda (s1)
                                    (cond [(null? s1) '()]
                                          [(member? (car s1) s2) (J (cdr s1))]
                                          [else (cons (car s1)
                                                      (J (cdr s1)))]))))
                        (cond [(null? s2) (hop '())]
                              [else (J s1)])))))
              (cond [(null? lset) '()]
                    [else (A lset)])))))

(module+ test
  (check-equal? (intersectall3 '((a b c) (b c) (b)))
                '(b)))

(define rember
  (lambda (a lat)
    (letrec ((R (lambda (lat)
                  (cond [(null? lat) '()]
                        [(equal? (car lat) a) (cdr lat)]
                        [else (cons (car lat)
                                    (R (cdr lat)))]))))
      (R lat))))

(module+ test
  (check-equal? (rember 'b '(a b c b d))
                '(a c b d)))

(define rember-beyond-first
  (lambda (a lat)
    (letrec ((R (lambda (lat)
                  (cond [(null? lat) '()]
                        [(eq? (car lat) a) '()]
                        [else (cons (car lat)
                                    (R (cdr lat)))]))))
      (R lat))))

(module+ test
  (check-equal? (rember-beyond-first 'd '(a b c d e f g))
                '(a b c)))

(define rember-upto-last1
  (lambda (a orig-lat)
    (letrec ((R (lambda (lat)
                  (cond [(null? lat) orig-lat]
                        [(eq? (car lat) a) (cdr lat)]
                        [else (R (cdr lat))]))))
      (R orig-lat))))

(module+ test
  (check-equal? (rember-upto-last1 'd '(a b c d e f g))
                '(e f g))
  (check-equal? (rember-upto-last1 'z '(a b c d e f g))
                '(a b c d e f g)))

(define rember-upto-last2
  (lambda (a lat)
    (let/cc
     skip
     (letrec ((R (lambda (lat)
                   (cond [(null? lat) '()]
                         [(eq? (car lat) a) (skip (R (cdr lat)))]
                         [else (cons (car lat)
                                     (R (cdr lat)))]))))
       (R lat)))))

(module+ test
  (check-equal? (rember-upto-last2 'd '(a b c d e f g))
                '(e f g))
  (check-equal? (rember-upto-last2 'z '(a b c d e f g))
                '(a b c d e f g)))
```

**Comment by greghendershott** [comment]
Related discussion: Issue #51.

Currently racket-mode is using `errortrace`'s execute-counts annotations. By contrast, DrRacket and the `cover` package do their own thing. The current plan is to use cover when https://github.com/florence/cover/pull/53 is merged. (Otherwise, plan B is I'll dig into the annotations myself. But Florence has invested a lot of time into this already so I'd like to leverage that if possible.)

**Comment by greghendershott** [comment]
A macro usage can expand to an `if`. 

For example `(check-true #t)` expands (partially) to:

```
(let-values (((/tmp/foo2.rkt:3:0) check-true1415117)
             ((temp1) '#t)
             ((temp2) '(#<path:/tmp/foo2.rkt> 3 0 33 15))
             ((temp3) '(check-true #t)))
  (if (#%app variable-reference-constant?
             (#%variable-reference check-true1415117))
    (#%app check-true1415115 temp3 '#t temp2 '#t temp1 '#f '#f)
    (#%app (#%app checked-procedure-check-and-extract
                  struct:keyword-procedure
                  /tmp/foo2.rkt:3:0
                  keyword-procedure-extract
                  '(#:expression #:location)
                  '3)
           '(#:expression #:location)
           (#%app list temp3 temp2)
           temp1)))
```

`errrortrace/stacktrace` instruments the entire `if` as well as each branch. All share the same source location -- the macro usage.  Therefore the coverage data may have multiple data points for the same source location. They could have a mix of `#t` and `#f` values, if just one arm is run.

Even if just one arm is taken, conceptually that chunk of source _is_ covered. The `check-true` macro _was_ run, and that bit of the original source _is_ covered. The implementation of the macro is N/A.

This works out in DrR because it effectively `or`s the coverage data for the same source location -- if any are `#t`, the location is covered. Whereas I was effectively `and`ing -- because I only care about highlighting uncovered stuff, I was filtering for the `#f`s. I'll fix that.

**Comment by greghendershott** [comment]
@zenspider After fixing that, there remain a few differences with the DrR output. For example:

```
(define rember
  (lambda (a lat)
    (letrec ((R (lambda (lat)
                  (cond [(null? lat) '()] ;; <--- DrR says the '() is not covered
                        [(equal? (car lat) a) (cdr lat)]
                        [else (cons (car lat)
                                    (R (cdr lat)))]))))
      (R lat))))

(module+ test
  (check-equal? (rember 'b '(a b c b d))
                '(a c b d))
```

I'm showing `rember` as fully covered, whereas DrR says that little `'()` fragment is _not_. Which do you think is the correct answer?

[I've also been working on instrumenting things the way DrR does -- it uses `errortrace/stacktrace` but not `errortrace-lib`. So I'm not surprised that DrR is giving a slightly different answer here, and I can get that different answer. But I don't yet understand why it's different. I'll think through that. But also I'm interested what you think, as the writer of the code.]

**Comment by greghendershott** [comment]
OK I looked at it more. DrR is correct. That cond branch isn't run.

The issue is with `'()`. Apparently that isn't instrumented for execution counts; it's "too primitive". So it can't be checked for a count of zero. (Same for `(list)`. Whereas something like `(apply list null)` does get instrumented and does get highlighted as uncovered.)

So I think the answer is I should keep working on the branch where I'm using the DrR approach to coverage (and other) instrumenting. Although that branch has an unrelated problem I'm still working on, the coverage is working. I'd already done the equivalent of commit 6c4f13c7a6 on that branch. I wanted to see if that could be a quick fix for the execute counts approach. Although not sufficient, it's a big improvement. I could merge it to master as an interim fix if the more-thorough branch is taking too long to finish fully.

tl;dr It's not done but I'm closer to understanding what the hell I'm doing. :)

**Comment by greghendershott** [comment]
That branch is https://github.com/greghendershott/racket-mode/tree/instrument. I scavenged elements from both DrRacket and errortrace-lib. Not to jinx it, but I _think_ I have it figured out now. Will test some more. Currently I expect to merge this full solution rather than the interim "mostly" fix I mentioned above.
